### DISCUSSION TITLE
Enforcing Prediction Intervals in Time Series Models

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1814] [FAQs]
# Background

In many time series problems, it is often more important to get a prediction interval for the forecasts, than to get a point estimate itself. For example, in the supply chain world, there is a concept of `cycle time` which is the time taken from the start of manufacturing to when the product is available to the end customer. In some cases, the cycle time can span several months. In these scenarios, it is important to forecast the customer demand ahead of time. One might also be interested in knowing how much manufacturing to plan for in a typical scenario (point estimate) as well as in the best and worst case scenarios (upper and lower prediction intervals). Luckily, the `pycaret` time series module allows the users to do this with ease. Let's see how it works.

NOTE: A complete notebook example can be **[found here](https://gist.github.com/ngupta23/12dc80fc39f0a57ac9b7416953114ece)**.

# Solution in `pycaret`

```python
#### Load data ----
from pycaret.datasets import get_data
y = get_data("airline")
```

##  Enforcing prediction intervals 
```python
#### Setup experiment ----
from pycaret.internal.pycaret_experiment.time_series_experiment import TimeSeriesExperiment
exp = TimeSeriesExperiment()

#### Enforce usage of only those models that have a prediction interval ----
exp.setup(data=y, fh=12, enforce_pi=True, session_id=42)

# Available Models (only those having ability to generate prediction intervals) ----
exp.models()
```
Typically ~ 30 models are available in `pycaret`. When `enforce_pi` is set to `True`, model building is limited to only those models that support prediction intervals. This is applicable to individual model building as well as while comparing models. 

![image](https://user-images.githubusercontent.com/33585645/140648276-0d2250a2-9d0e-4f53-8d22-9318f4f753e0.png)

### Creating models & making predictions with intervals

```python
#### Create model & plot predictions ----
model = exp.create_model("ets")
exp.plot_model(model)
```
**NOTE: By default, (95%) prediction interval will be plotted for models that support it. Also note that the plots are interactive and by hovering over them, users can see the values.**

![image](https://user-images.githubusercontent.com/33585645/140648375-b32099f2-baa2-45e2-803a-81a945a70d5e.png)

```python
#### Get Predictions with prediction intervals (Default: 95%) ----
exp.predict_model(model, return_pred_int=True)
```

![image](https://user-images.githubusercontent.com/33585645/140648341-fdcd8474-26f0-45a0-a88a-4b6534e97a51.png)

### Comparing multiple models
```python
best = exp.compare_models(turbo=False)
```

![image](https://user-images.githubusercontent.com/33585645/140648437-be3bea5d-238b-4247-a2da-e23e0ef5d7ab.png)

NOTE: Again, the comparison is limited to only those models that have a prediction interval.

That is how simple it is to build and compare models that support prediction intervals with `pycaret`. That is it for this FAQ. Be sure to read the suggested next reads.

Happy Forecasting!

# Suggested Next Reads
- https://github.com/pycaret/pycaret/discussions/1757
- https://github.com/pycaret/pycaret/discussions/1808

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
